### DISCUSSION TITLE
Does uniforms SchemaBridge supports the String Built-in Formats?

### DISCUSSION THREAD

**Discussion by rafaeltuelho** [vazco/uniforms#941] [Q&A] [answered]
It seems the `JSONSchemaBridge` does not support the string built-in formats [1] and [2]
Let's say my json schema define a property type as string but its format as `date`...
```json
"Date": {
  "format": "date",
  "type": "string",
}
```

How can I tell uniforms to render that property as Date and not Text?

[1] https://json-schema.org/understanding-json-schema/reference/string.html#built-in-formats
[2] https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.3

**Comment by rafaeltuelho** [comment]
I see `JSONSchemaBridge` recognizes only `date-time` as field format
https://github.com/vazco/uniforms/blob/master/packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts#L330

**Comment by radekmie** [comment, marked as answer]
Hi, @rafaeltuelho! As you already noticed, it does support `date-time`. The problem with date/time/datetime inputs is that there's no good HTML support across all browsers (Firefox and Safari were "behind"). However, the goal of uniforms was to provide themes that can handle _all_ schemas, at least at a decent level (mind the difference: all of them as good as possible, not as many as possible ideally).

In the end, we ended up with `datetime-local` ([source](https://github.com/vazco/uniforms/blob/7f1fddbf9dea9572835536d321fe4098de60e355/packages/uniforms-unstyled/src/DateField.tsx#L48)) to make it unified across themes. It's not ideal, but (mostly) works.

There are **a lot** of date and time pickers out there, and so we suggest creating custom fields using them, based on your theme and preference. And to make it "automatic", i.e., work with `AutoField`, all you need to do is to tell `AutoField` to use your custom field if the `format` is `date` (or anything else). It's [here](https://uniforms.tools/docs/uth-autofield-algorithm#overriding-autofield) in the docs.

Having said all that... Safari 14.1 got released literally this week and supports all of these ([link](https://webkit.org/blog/11648/new-webkit-features-in-safari-14-1/)). Finally!

**Reply by rafaeltuelho** [reply]
Thanks @radekmie for this detailed explanation. I'll take a look at the `AutoField` approach. I was wondering if is there a way to do this at the `AutoForm` level as I'm trying to make my form rendering as generic as possible.

**Reply by radekmie** [reply]
You can do it even higher in the JSX tree - this `AutoField` configuration provider can be higher than any form.

**Reply by rafaeltuelho** [reply]
I see. I'm trying overriding the `Autofield` but I'm getting this error

![image](https://user-images.githubusercontent.com/39063/116697948-cbd4b900-a991-11eb-94e0-33a36e3acb18.png)

It seems it can't access the `Provider` 🤔

**Reply by radekmie** [reply]
What version of uniforms are you using? It got released in [3.1.0](https://github.com/vazco/uniforms/blob/master/CHANGELOG.md#v310-2021-02-03).

**Reply by rafaeltuelho** [reply]
```json
    "uniforms": "^3.4.0",
    "uniforms-bridge-json-schema": "^3.4.0",
    "uniforms-bridge-simple-schema": "^3.4.0",
    "uniforms-patternfly": "^4.2.1",
```

**Reply by rafaeltuelho** [reply]
In this meanwhile, I'm trying to patch my raw JsonSchema (that I'm getting from an OpenAPI (swagger) document) to replace `format: 'date'` to `format: 'date-time'` before I pass it to the JsonSchemaBridge.

**Reply by radekmie** [reply]
Well, it's something that `uniforms-patternfly` have to catch up with - [the `AutoField` of this theme](https://github.com/aerogear/uniforms-patternfly/blob/29b864d2f504879db39ebe5a2a36edb28c7d56cd/src/AutoField.tsx#L22) is not using the new `createAutoField` helper as [official themes do](https://github.com/vazco/uniforms/blob/7f1fddbf9dea9572835536d321fe4098de60e355/packages/uniforms-unstyled/src/AutoField.tsx#L15).

**Reply by rafaeltuelho** [reply]
Thanks to pointing that out @radekmie

**Reply by wtrocki** [reply]
@radekmie we will update this.

Thank you so much
